core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1646,7 +1674,7 @@ private static List<HistoryEntry> updateSnapshotLog(
       List<HistoryEntry> newSnapshotLog = Lists.newArrayList();
       for (HistoryEntry logEntry : snapshotLog) {
         long snapshotId = logEntry.snapshotId();
-        if (snapshotsById.containsKey(snapshotId)) {
+        if (snapshotsById.containsKey(snapshotId) || lazySnapshotLoadingEnabled) {
```

Comment:
Right now, I think all we need to do is detect whether we need to alter the snapshot history at all:

```java
      boolean hasRemovals = changes(MetadataUpdate.RemoveSnapshot.class).findAny().isPresent();
```

That will tell us whether any snapshots were removed from history and whether we need to clear history when a snapshot is missing. If there aren't removals, then we should use this branch.

Alternatively, you could detect whether the snapshot supplier is non-null because that indicates that some snapshots are missing. But I'm concerned that if we do that, then when we fix the current behavior that loads all snapshots at write time we will potentially keep extra history that doesn't correspond to a snapshot. I'd rather only keep history if we know it is safe to do so because there weren't any removed snapshots.

Reply:
In my initial version of the code I had almost the same logic, but it was a bit difficult to understand because we basically need to handle the snapshot log handling both for the client and the server side.
I'm also not a fan of exposing `removeLazilyLoadedSnapshotsIf()`, but it makes the intent a bit clearer.

We've also talked with @danielcweeks  to explore lazy metadata loading in general and I've opened https://github.com/apache/iceberg/pull/8173. This would allow solving us this particular issue but also allow the server to send a "slim" version of `TableMetadata` and load remaining data when requested. In particular, snapshots / history / metadata log can become quite long.

Reply:
> In my initial version of the code I had almost the same logic, but it was a bit difficult to understand because we basically need to handle the snapshot log handling both for the client and the server side.

I'm not following why this is a problem. On the server side there's no `RemoveSnapshot` in changes, so the history is sent as-is. On the client side the same would happen most of the time.

Reply:
In fact there are `RemoveSnapshot` entries on the server-side when `TableMetadata#removeSnapshotsIf(..)` is getting called, which can be seen below:
![image](https://github.com/apache/iceberg/assets/271029/c00d9670-14b2-4ee7-b790-cc025fdaef90)